mpisppy/opt/presolve.py

```diff
@@ -0,0 +1,247 @@
+                        _lb_generator(node.nonant_vardata_list),
+                        dtype=float,
+                        count=nlen,
+                    )
+                    if ndn in local_lower_bounds:
+                        np.maximum(
+                            local_lower_bounds[ndn],
+                            scenario_node_lower_bounds,
+                            out=local_lower_bounds[ndn],
+                        )
+                        if same_nonant_bounds:
+                            same_nonant_bounds = np.allclose(
+                                local_lower_bounds[ndn], scenario_node_lower_bounds
+                            )
+                    else:
+                        local_lower_bounds[ndn] = scenario_node_lower_bounds
+                        global_lower_bounds[ndn] = np.zeros(nlen, dtype=float)
+
+                    # gather upper bounds
+                    scenario_node_upper_bounds = np.fromiter(
+                        _ub_generator(node.nonant_vardata_list),
+                        dtype=float,
+                        count=nlen,
+                    )
+                    if ndn in local_upper_bounds:
+                        np.minimum(
+                            local_upper_bounds[ndn],
+                            scenario_node_upper_bounds,
+                            out=local_upper_bounds[ndn],
+                        )
+                        if same_nonant_bounds:
+                            same_nonant_bounds = np.allclose(
+                                local_upper_bounds[ndn], scenario_node_upper_bounds
+                            )
+                    else:
+                        local_upper_bounds[ndn] = scenario_node_upper_bounds
+                        global_upper_bounds[ndn] = np.zeros(nlen, dtype=float)
+
+            # reduce lower bounds
+            for ndn, local_bounds in local_lower_bounds.items():
+                self.opt.comms[ndn].Allreduce(
+                    [local_bounds, MPI.DOUBLE],
+                    [global_lower_bounds[ndn], MPI.DOUBLE],
+                    op=MPI.MAX,
+                )
+                if same_nonant_bounds:
+                    same_nonant_bounds = np.allclose(
+                        global_lower_bounds[ndn], local_bounds
+                    )
+
+            # reduce upper bounds
+            for ndn, local_bounds in local_upper_bounds.items():
+                self.opt.comms[ndn].Allreduce(
+                    [local_bounds, MPI.DOUBLE],
+                    [global_upper_bounds[ndn], MPI.DOUBLE],
+                    op=MPI.MIN,
+                )
+                if same_nonant_bounds:
+                    same_nonant_bounds = np.allclose(
+                        global_upper_bounds[ndn], local_bounds
+                    )
+
+            # At this point, we've either proved that
+            # there are tighter bounds or not.
+            # If not, we can quit
```

Comment:
I'm not sure this is precisely true, although I think I agree with the conclusion. It really depends on the stopping criteria for FBBT (improvement_tol and max_iter). It is technically possible to run FBBT twice and have the second iteration do something useful, even without any changes to the model in between. This is most likely when max_iter is small. Maybe just update the comment to say something like "the nonant bounds did not change, so it is unlikely that more FBBT will do anything"???

Reply:
The default `max_iter` is relatively large, so this shouldn't be a problem. However, if you ever allow users to configure the `IntervalTightener`s for the subproblems, it could be an issue.